File: ProgrammingProjects/cpp/Project_02/main.cpp
#include "chemistry/vibrational_analysis.h"

#include <chemistry/molecule.h>

#include <exception>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <stdexcept>

int main(int argc, char* argv[])
{
    if (argc != 3) {
        std::cerr
            << "Usage: project02 <geometry-file> <hessian-file>\n";
        return 1;
    }

    try {
        const std::filesystem::path geometry_path{argv[1]};
        const std::filesystem::path hessian_path{argv[2]};

        const chemistry::Molecule molecule =
            chemistry::Molecule::from_file(geometry_path);

        std::ifstream hessian_file{hessian_path};
        if (!hessian_file) {
            throw std::runtime_error(
                "Cannot open Hessian file: " +
                hessian_path.string()
            );
        }

        std::cout
            << "Number of atoms: "
            << molecule.atom_count() << '\n'
            << "Hessian dimension: "
            << 3 * molecule.atom_count() << " x "
            << 3 * molecule.atom_count() << '\n';

        // Step 2: Read the Cartesian Hessian into an Eigen::MatrixXd.
        chemistry::VibrationalAnalysis vib = chemistry::VibrationalAnalysis{molecule, hessian_path};
        const Eigen::MatrixXd& hessian = vib.hessian();
        std::cout << "Hessian:\n" << hessian << '\n';

        // Step3: 获得质权坐标下的Hessian
        Eigen::MatrixXd F = vib.get_mass_weighted_hessian();
        std::cout <<"mass weighted Hessian:\n" << F << '\n';

        // Step4: 对角化F，从而得到特征值和normal mode
        const chemistry::NormalModeResult result = vib.get_normal_modes();

        std::cout
            << "Eigenvalues:\n"
            << result.eigenvalues
            << '\n';

        std::cout
            << "Normal modes:\n"
            << result.normal_modes
            << '\n';

        // Step5: 转换特征值的单位，得到以cm-1为单位的谐振频率
        const Eigen::VectorXd wavenumbers = vib.get_wavenumbers_cm_inverse();

        std::cout << "Wavenumbers (cm^-1):\n"
                  << wavenumbers << '\n';

    }
    catch (const std::exception& error) {
        std::cerr << "Error: " << error.what() << '\n';
        return 1;
    }

    return 0;
}

File: ProgrammingProjects/python/Project_02/main.py
from pathlib import Path

from chemistry import Molecule
from chemistry.vibrational_analysis import VibrationalAnalysis

INPUT_DIRECTORY = (
    Path(__file__).resolve().parents[2]
    / "resources"
    / "Project_02"
    / "input"
)

GEOMETRY_FILE = INPUT_DIRECTORY / "h2o_geom.txt"
HESSIAN_FILE = INPUT_DIRECTORY / "h2o_hessian.txt"


def main(
    geometry_file: Path = GEOMETRY_FILE,
    hessian_file: Path = HESSIAN_FILE,
) -> None:
    molecule = Molecule.from_file(geometry_file)

    if not hessian_file.is_file():
        raise ValueError(
            "Cannot open Hessian file: "
            f"{hessian_file}"
        )

    dimension = 3 * molecule.atom_count
    print(f"Number of atoms: {molecule.atom_count}")
    print(f"Hessian dimension: {dimension} x {dimension}")

    # Step 2: Read the Cartesian Hessian into a NumPy array.
    vibrational_analysis = VibrationalAnalysis(molecule, hessian_file)
    print('hessian:')
    print(vibrational_analysis.hessian)

    # Step 3: 获取质权坐标下的Hessian
    mass_weighted_hessian = vibrational_analysis.get_mass_weighted_hessian()
    print('mass_weighted_hessian:')
    print(mass_weighted_hessian)

    # Step 4: 对角化F，从而得到特征值和normal_mode
    eigen_values, normal_modes = vibrational_analysis.get_normal_modes()
    print('eigen_values:')
    print(eigen_values)
    print('normal_modes:')
    print(normal_modes)
    # Step 5: Convert eigenvalues to spectroscopic wavenumbers.
    wavenumbers = (
        vibrational_analysis.get_wavenumbers_cm_inverse()
    )
    print("wavenumbers (cm^-1):")
    for mode_index, wavenumber in enumerate(wavenumbers):
        print(f"{mode_index:3d} {wavenumber:14.6f}")

if __name__ == "__main__":
    main()
